docs/content/rules/sort-modules.mdx

```diff
@@ -0,0 +1,592 @@
+    'unknown',
+   ],
++  customGroups: [                               // [!code ++]
++    [                                           // [!code ++]
++      {                                         // [!code ++]
++         groupName: 'input-types-and-interfaces', // [!code ++]
++         anyOf: [                               // [!code ++]
++           {                                    // [!code ++]
++              selector: 'type',                 // [!code ++]
++              elementNamePattern: 'Input'.      // [!code ++]
++           },                                   // [!code ++]
++           {                                    // [!code ++]
++              selector: 'interface',            // [!code ++]
++              elementNamePattern: 'Input'.      // [!code ++]
++           },                                   // [!code ++]
++         ]                                      // [!code ++]
++      },                                        // [!code ++]
++      {                                         // [!code ++]
++         groupName: 'output-types-and-interfaces', // [!code ++]
++         anyOf: [                               // [!code ++]
++           {                                    // [!code ++]
++              selector: 'type',                 // [!code ++]
++              elementNamePattern: 'Output'      // [!code ++]
++           },                                   // [!code ++]
++           {                                    // [!code ++]
++              selector: 'interface',            // [!code ++]
++              elementNamePattern: 'Output'      // [!code ++]
++           },                                   // [!code ++]
++         ]                                      // [!code ++]
++      },                                        // [!code ++]
++      {                                         // [!code ++]
++         groupName: 'unsorted-functions',       // [!code ++]
++         type: 'unsorted',                      // [!code ++]
++         selector: 'function',                  // [!code ++]
++      },                                        // [!code ++]
++    ]                                           // [!code ++]
++  ]                                             // [!code ++]
+ }
+```
+
+
+## Usage
+
+<CodeTabs
+  code={[
+    {
+      source: dedent`
+        // eslint.config.js
+        import perfectionist from 'eslint-plugin-perfectionist'
+
+        export default [
+          {
+            plugins: {
+              perfectionist,
+            },
+            rules: {
+              'perfectionist/sort-modules': [
+                'error',
+                {
+                  type: 'alphabetical',
+                  order: 'asc',
+                  ignoreCase: true,
+                  specialCharacters: 'keep',
+                  partitionByComment: false,
+                  partitionByNewLine: false,
+                  newlinesBetween: 'ignore',
+                  groups: [
```

Comment:
To open a discussion, maybe I would propose the following order:

```
'declare-enum',
'enum',
'export-enum',
['declare-interface', 'declare-type'],
['interface', 'type'],
['export-interface', 'export-type'],
'declare-class',
'class',
'export-class',
'declare-function',
'function'
'export-function',
```

because it seems more logical, e.g. to avoid dependency issues like this:
```
export interface A {}

interface B {}

export interface C {}
```

while all the exports grouped at the end would sort without care about such dependencies:

```
interface B {}

export interface A {}

export interface C {}
```

Reply:
@OlivierZal `interfaces` and `types` do not suffer dependency issues like `classes` and `enums` have (they are eliminated at runtime and typescript does not require interfaces or types to be ordered for compilation).

The question is valid for `class` and `enum` though, and I think that we have two schools:
- Put `export-ed` elements at the end to ensure that no dependency issues occur (what you propose).
- Put `export-ed` elements first (they are generally above in the codebases I have seen), and rely on our dependency detection system to place `class`/`enum` above the exported element if needed.

I'm personally for the second one: I trust our dependency detection system (even though it might not be bulletproof at first, `sort-class` is now very reliable on that matter). But it's really a matter of personal preference in the end.

Reply:
Functions on the other hand don't depend on other functions (or at least, it's not required at runtime to have them ordered), so I propose to do a bit of both schools:

- Place `type`, `interface`, `class` exports at the end.
- Place `export function` first (Industry standard?)
- What about enums?

Reply:
Yes, take `interface` as an example was not smart 😅 but you got my point, and I've no strong option since default can be overridden with a custom configuration according to user's preferences.

So I'm also ok with option 2, it was just to open a discussion.

Reply:
Your argument regarding interfaces/types/classes having sub-types/interfaces/classes makes a lot of sense, so I'm actually more incline toward the hybrid proposal than option 2 now!

Reply:
I'll try to search for official guidelines on this and see what comes out.

Reply:
I'm wondering if we're not missing a use-case:
A generic `types.ts` file exporting

```ts
export type A = {
    attribute1: D
}
type D = "d"

export type B = {
    attribute1: C
}
type C = "c"
``` 

The sorted result is
```ts
export type A = {
    attribute1: D
}
export type B = {
    attribute1: C
}

type C = "c"
type D = "d"
```

I think that some users would prefer to have the 1st sort: it's essentially the same as if we activated the dependency system for `types` (or `interfaces`), but reversed: dependencies are put right after their `parent` rather than before.

Note: if we allow users to enable this (through some sort of `takeTypesDependenciesIntoAccount` option), it's important to also allow them to choose what types of dependencies to take into account:

```ts
export MyEnum {}

export type A = {
    attribute1: MyEnum
}
```

Users might not want to put `MyEnum` after `A`.

I was thinking of an API like this in options:

```ts
{
  typeDependencies: "ignore" | { position: "before" | "after"; types: ('type' | 'interface' | 'enum' | 'class')[]
}
```

Reply:
Yes, could be postponed. 

I like your suggestion (ignore, before, after), but I'm not sure to see why options should separate options for types and interfaces.

Reply:
I identify multiple `export` use cases (with or without `type`):

```
export = App

export { App }

export default App

export { default as App }

export { App } from './some-import.js'
```

Reply:
I've updated the default groups order to place `export-class` after `class`. It looks like placing exported interfaces and types first seem standard in the codebases I have seen though. I'm sure we are going to be improving this over time, but there are too many variables to take into account that I don't objectively know what would be the best approach to take regarding this, so I think we can wait for some user feedback on that.